### PR TITLE
Remove custom _c10d_functional_autograd implementations, use redirects

### PR DESCRIPTION
**Pull request by vishal9-team**
Stack from [ghstack](https://github.com/ezyang/ghstack/tree/0.15.0) (oldest at bottom):
* __->__ #172792

In #168140 support was added to functional collectives in _c10d_functional.* to support gradient calculation. Partial support for the same existed in _c10d_functional_autograd.*

This PR normalizes support for autograd in functional collectives, by 
[a] removing C++ implementations in _c10d_functional_autograd.*
[b] re-routing python .*_autograd counterparts to use standard functional collectives instead

Note: [b] is for backward compatibility and can be removed at sometime in the future

Previously

`all_gather_tensor_autograd -> _c10d_functional_autograd.all_gather_into_tensor -> all_gather_into_tensor_autograd
`
Now,

`all_gather_tensor_autograd -> all_gather_into_tensor -> _c10d_functional.all_gather_into_tensor -> all_gather_into_tensor`

Also,

`_c10d_functional_autograd.all_gather_into_tensor -> all_gather_into_tensor
`
Removed
C++ `all_gather_into_tensor_autograd`

### CODE REVIEW

**Comment by wwwjn** [comment]
I want to provide a real use case why we need this: https://github.com/pytorch/torchtitan/pull/3142#discussion_r3193185538 . @vishal9-team  Do you plan to follow up on this?

**Comment by vishal9-team** [comment]
> I want to provide a real use case why we need this: https://github.com/pytorch/torchtitan/pull/3142#discussion_r3193185538 . @vishal9-team  Do you plan to follow up on this?
> 

Sorry what do you mean by 'this'? The PR?

It's actually ready for shipping, I was waiting for Will's review. But if you want to approve, I can rebase and merge

**Comment by wwwjn** [comment]
> Sorry what do you mean by 'this'? The PR?
>

Yes, for this PR.

> It's actually ready for shipping, I was waiting for Will's review. But if you want to approve, I can rebase and merge

@wconstab  Can you help review and we can go ahead and merge?

File: test/distributed/test_functional_differentials.py [hunk 1]
```diff
@@ -597,6 +597,142 @@ def test_all_to_all_single_opcheck(self, test_utils):
             test_utils=test_utils,
         )
 
+    # ============================================================
+    # _c10d_functional_autograd Backward Compatibility Tests
+    # ============================================================
+
+    @parametrize("device", devices)
+    def test_all_gather_into_tensor_autograd_backward(self, device):
```
Comment (wconstab) [inline test/distributed/test_functional_differentials.py:601]: a few nits about these tests

1) they could be easier to understand, the comment mentions 'autograd ops' and 'c10d_functional' ops.  The code has 'c10d_functional_autograd' ops and 'funcol' ops.  Prefer the explanation and the code (var names) matched more closely
2) the world size isn't asserted in the test, so doing expected grad = 3,3 is a bit risky imo, better to be explicit
3) the tests talk about comparing one function to the other, but they actually compare just the one function to a reference. maybe the tests could be more easily parametrized and covered in one test case if  they looped over 3 functions and just compared the a and b version of each function to each other.

File: test/distributed/test_functional_differentials.py [hunk 2]
```diff
@@ -597,6 +597,142 @@ def test_all_to_all_single_opcheck(self, test_utils):
...
+
+    @parametrize("test_utils", autograd_test_utils)
+    def test_all_gather_into_tensor_autograd_opcheck(self, test_utils):
+        """Test _c10d_functional_autograd.all_gather_into_tensor op registration."""
+        group_name = dist.group.WORLD.group_name
+
+        input_tensor = torch.ones(3, 3, requires_grad=True)
+
+        torch.library.opcheck(
+            torch.ops._c10d_functional_autograd.all_gather_into_tensor,
```
Comment (wconstab) [inline test/distributed/test_functional_differentials.py:698]: what does this actually test- just that the op is actually registered? is it implicitly covered by having another test above that calls this op?

File: torch/distributed/_functional_collectives.py [hunk 3]
```diff
@@ -1224,24 +1197,6 @@ def _resolve_group_name(group: RANK_TYPES, tag: str = "") -> c10d.GroupName:
         raise ValueError(f"Unsupported group type: {type(group)}, {group}")
 
 
-class _FromTorchTensor(torch.autograd.Function):
```
Comment (wconstab) [inline torch/distributed/_functional_collectives.py:1367]: i'm not sure why this part was removed, was it only needed for the c10d_functional_autograd variants and not for the c10d_functional variants with autograd?  If so, can you explain why? Should I expect test coverage to notice this change?

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 4]
```diff
@@ -678,6 +472,40 @@ at::Tensor shard_dim_alltoall(
...
+// separately in Python via torch.library.register_autograd.
+TORCH_LIBRARY(_c10d_functional_autograd, m) {
+  m.def(
+      "all_to_all_single("
+      "Tensor input, "
+      "SymInt[] output_split_sizes, "
+      "SymInt[] input_split_sizes, "
+      "str group_name) -> Tensor",
+      torch::dispatch(
+          c10::DispatchKey::CompositeExplicitAutograd, c10d::all_to_all_single),
```
Comment (wconstab) [inline torch/csrc/distributed/c10d/Functional.cpp]: it would be nice if there were a code note in one place that described the whole design of the _c10d_functional_autograd BC layer.  it seems to have CIA keys registered here, and backward keys registered somewhere else.

Comment (vishal9-team) [reply torch/csrc/distributed/c10d/Functional.cpp]: I have this here - do you feel it's incomplete?

`// Backward compatibility: _c10d_functional_autograd namespace
// These ops are identical to _c10d_functional ops. Autograd is registered
// separately in Python via torch.library.register_autograd.`

File: torch/distributed/_functional_collectives.py (hunk 3 above)
Comment (vishal9-team) [reply torch/distributed/_functional_collectives.py:1367]: For BC, I'm keeping c10d_functional_autograd.*_autograd() calls but redirecting them to c10d_functional.*() calls in python

_FromTorchTensor is only used in .*_autograd() implementations which this PR removes and hence this is no longer needed

This change should be transparent to test coverage

File: test/distributed/test_functional_differentials.py (hunk 2 above)
Comment (vishal9-team) [reply test/distributed/test_functional_differentials.py:698]: As per discussions with Alban/Richard eager/opcheck/compile are overlapping, but not redundant and hence, I've included these for all functional differential tests

File: test/distributed/test_functional_differentials.py (hunk 1 above)
Comment (vishal9-team) [reply test/distributed/test_functional_differentials.py:601]: I agree "autograd ops" is a bit overloaded and should read ".*_autograd", I'm hoping we can eventually remove _c10d_functional_autograd.* and delete most of these

File: torch/distributed/_functional_collectives.py (hunk 3 above)
Comment (wconstab) [reply torch/distributed/_functional_collectives.py:1367]: ok, i missed that _FromTorchTensor was only used in the deleted code, that's fine then. thanks for the clarification.

File: torch/csrc/distributed/c10d/Functional.cpp (hunk 4 above)
Comment (wconstab) [reply torch/csrc/distributed/c10d/Functional.cpp]: i guess it just doesn't say "why" - i assume if you could have put both registrations in the same file you would have. If you can't, its nice to mention where the other one is so its easy for someone to find it when they are discovering the code.

**Review by wconstab — APPROVED** [review]
lgtm.

**Comment by wwwjn** [comment]
@vishal9-team  Could you help following up and merge the PR? Very appreciated!

File: test/distributed/test_functional_differentials.py [hunk 5]
```diff
@@ -593,6 +593,142 @@ def test_all_to_all_single_opcheck(self, test_utils):
             test_utils=test_utils,
         )
 
+    # ============================================================
+    # _c10d_functional_autograd Backward Compatibility Tests
+    # ============================================================
+
+    @parametrize("device", devices)
+    def test_all_gather_into_tensor_autograd_backward(self, device):
```
Comment (d4l3k) [inline test/distributed/test_functional_differentials.py:601]: should this use the new _single name or is it intentional that it uses the old functions for BC compatibility?

File: torch/csrc/distributed/c10d/Functional.cpp [hunk 6]
```diff
@@ -1130,6 +917,57 @@ at::Tensor shard_dim_alltoall(
...
+      "str reduce_op, "
+      "int group_size, "
+      "Any group_name) -> Tensor",
+      torch::dispatch(
+          c10::DispatchKey::CompositeExplicitAutograd,
+          [](const at::Tensor& input,
+             std::string reduce_op,
+             int64_t group_size,
+             const c10::IValue& group) {
+            return c10d::reduce_scatter_tensor(
```
Comment (d4l3k) [inline torch/csrc/distributed/c10d/Functional.cpp:946]: should this use the new names?

File: test/distributed/test_functional_differentials.py (hunk 5 above)
Comment (wconstab) [reply test/distributed/test_functional_differentials.py:601]: it seems like the intent was to test BC. (see comment above).

File: torch/csrc/distributed/c10d/Functional.cpp (hunk 6 above)
Comment (wconstab) [reply torch/csrc/distributed/c10d/Functional.cpp:946]: I don't think it can.

IIUC, the _single rename (from https://github.com/pytorch/pytorch/pull/186125) only applied to the Python wrappers in _functional_collectives.py; it didn't rename the C++ c10d:: functions or the registered op schemas.  Does that match your understanding or am I confused?

**Comment by wconstab** [comment]
@claude can you diagnose the torchtitan-test CI fail? is it caused by this PR or is it something upstream?
can you also review the PR in its current state, post rebase?
